### PR TITLE
Set proper socket parameters suitable for CLI communication

### PR DESCRIPTION
**Pull request by kohsuke**
Keep-alive option ensures that the connection won't hang forever. NoDelay option improves performance as the communication over the socket is already datagram oriented.

See: https://developer.cloudbees.com/bin/view/DEV/On-Premise+Executors

@reviewbybees

### CODE REVIEW

**Comment by jtnord** [comment]
:+1:

**Comment by jglick** [comment]
:+1: assuming tests pass

**Comment by kohsuke** [comment]
One test didn't pass, but I'm pretty sure this is unrelated...

File: cli/src/main/java/hudson/cli/CLI.java [hunk 1]
```diff
@@ -169,12 +169,18 @@ protected void onDead() {
 
     private Channel connectViaCliPort(URL jenkins, CliPort clip) throws IOException {
         LOGGER.log(FINE, "Trying to connect directly via TCP/IP to {0}", clip.endpoint);
-        final Socket s;
+        final Socket s = new Socket();
+        // this prevents a connection from silently terminated by the router in between or the other peer
+        // and that goes without unnoticed. However, the time out is often very long (for example 2 hours
+        // by default in Linux) that this alone is enough to prevent that.
```
Comment (daniel-beck) [inline cli/src/main/java/hudson/cli/CLI.java]: Is this comment really grammatical?

**Pull request merged**